datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,820 @@
+                let bucket = id_to_val.try_insert_no_grow(new_hash, mi);
+                let bucket = match bucket {
+                    Ok(bucket) => bucket,
+                    Err(new_item) => {
+                        // this should basically never happen, but if it does, we must rebuild
+                        let bucket = id_to_val.insert(new_hash, new_item, |mi| mi.hash);
+                        val_to_idx.clear();
+                        unsafe {
+                            for bucket in id_to_val.iter() {
+                                let buk_idx = id_to_val.bucket_index(&bucket);
+                                let hi = HeapItem::new(
+                                    bucket.as_ref().val,
+                                    buk_idx,
+                                    self.sort,
+                                );
+                                val_to_idx.insert(hi);
+                            }
+                        }
+                        bucket
+                    }
+                };
+                let bucket_idx = unsafe { id_to_val.bucket_index(&bucket) };
+                let hi = HeapItem::new(new_val, bucket_idx, self.sort);
+                val_to_idx.insert(hi);
+                return Ok(());
+            }
+            Some(bucket) => bucket,
+        };
+
+        // this is a value for an existing group
+        let existing_mi = unsafe { old_bucket.as_mut() };
+        if cmp(&self.sort, &new_val, &existing_mi.val) != Ordering::Less {
+            // worse than the existing value _for this group_
+            return Ok(());
+        }
+
+        // update heap
+        let bucket_idx = unsafe { id_to_val.bucket_index(&old_bucket) };
+        let target_hi = HeapItem::new(existing_mi.val, bucket_idx, self.sort);
+        assert!(val_to_idx.remove(&target_hi), "Missing entry!");
+
+        let hi = HeapItem::new(new_val, bucket_idx, self.sort);
+        val_to_idx.insert(hi);
+
+        // update map
+        existing_mi.val = new_val;
+
+        Ok(())
+    }
+
+    pub fn len(&self) -> usize {
+        self.id_to_val.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.id_to_val.is_empty()
+    }
+
+    pub fn is_full(&self) -> bool {
+        self.len() >= self.limit
+    }
+
+    pub fn worst_val(&mut self) -> Option<&Option<VAL>> {
+        if !self.is_full() {
+            return None;
+        }
+        let val = self.val_to_idx.last()?;
+        Some(&val.val)
+    }
+
+    pub fn drain(&mut self) -> Vec<(ID, Option<VAL>)> {
+        let mut res = Vec::with_capacity(self.id_to_val.len());
+        while let Some(entry) = self.val_to_idx.pop_first() {
+            let bucket = unsafe { self.id_to_val.bucket(entry.buk_idx) };
+            let (bucket, _) = unsafe { self.id_to_val.remove(bucket) };
+            res.push((bucket.id, entry.val));
+        }
+        res
+    }
+}
+
+pub fn cmp<VAL: ArrowNativeTypeOp>(
+    sort: &SortOptions,
+    left: &Option<VAL>,
```

Comment:
I think `Option<&VAL>` is usually what you want.

Reply:
(as `Option` itself can be dereferenced while `VAL` might not).

Reply:
I went with this because now that we store nullable `VAL`s, the signature is:

```
struct HeapItem<VAL: ValueType> {
    val: Option<VAL>,
    buk_idx: usize,
    sort: SortOptions,
}
```

So any time I want to compare two of these, I already have a `Option<VAL>` to work with. But I feel like maybe I don't understand what you are saying?

Reply:
What I am saying is that you don't have to pass `Option` by reference.

This compiles:
```
index 9573edbbc..648a3dc4b 100644
--- a/datafusion/core/src/physical_plan/aggregates/priority_map.rs
+++ b/datafusion/core/src/physical_plan/aggregates/priority_map.rs
@@ -181,7 +181,7 @@ where
                 // Check goes here, because it is generalizable between str/String and Row/OwnedRow
                 let id = ids.value(row_idx);
                 if let Some(worst) = self.priority_map.worst_val() {
-                    if cmp(&sort, &val, worst) == Ordering::Greater {
+                    if cmp(&sort, val, *worst) == Ordering::Greater {
                         continue;
                     }
                 }
@@ -338,7 +338,7 @@ impl<VAL: ValueType> PartialOrd<Self> for HeapItem<VAL> {
 
 impl<VAL: ValueType> Ord for HeapItem<VAL> {
     fn cmp(&self, other: &Self) -> Ordering {
-        let res = cmp(&self.sort, &self.val, &other.val);
+        let res = cmp(&self.sort, self.val, other.val);
         if res != Ordering::Equal {
             return res;
         }
@@ -368,7 +368,7 @@ where
         // if we're full, and the new val is worse than all our values, just bail
         if is_full {
             let worst_entry = val_to_idx.last().expect("Missing value!");
-            if cmp(&self.sort, &new_val, &worst_entry.val) == Ordering::Greater {
+            if cmp(&self.sort, new_val, worst_entry.val) == Ordering::Greater {
                 return Ok(());
             }
         }
@@ -418,7 +418,7 @@ where
 
         // this is a value for an existing group
         let existing_mi = unsafe { old_bucket.as_mut() };
-        if cmp(&self.sort, &new_val, &existing_mi.val) != Ordering::Less {
+        if cmp(&self.sort, new_val, existing_mi.val) != Ordering::Less {
             // worse than the existing value _for this group_
             return Ok(());
         }
@@ -470,8 +470,8 @@ where
 
 pub fn cmp<VAL: ArrowNativeTypeOp>(
     sort: &SortOptions,
-    left: &Option<VAL>,
-    right: &Option<VAL>,
+    left: Option<VAL>,
+    right: Option<VAL>,
 ) -> Ordering {
     let (me, them) = match (left, right) {
         (None, None) => return Ordering::Equal,
@@ -499,7 +499,7 @@ pub fn cmp<VAL: ArrowNativeTypeOp>(
         }
         (Some(me), Some(them)) => (me, them),
     };
-    let mut res = me.compare(*them);
+    let mut res = me.compare(them);
     if sort.descending {
         res = res.reverse();
     }
```